Products/ZenModel/migrate/addZenHubWorkerService.py

```diff
@@ -0,0 +1,222 @@
+
+        # Make sure the zenhubworker service hasn't already been deployed
+        count = sum((1 for s in ctx.services if s.name == "zenhubworker"), 0)
+        if count > 0:
+            log.info("zenhubworker already deployed.  Skipping this step.")
+            return
+
+        # Locate ZenHub service definitions
+        services = [s for s in ctx.services if s.name == "zenhub"]
+        for service in services:
+
+            self._limitZenHubInstances(service)
+            self._removeObsoleteHealthchecks(service)
+            self._updateOriginalZenHubConfig(service)
+            workers, callLimit = self._updateZenHubConfig(service)
+
+            # Retrieve the zenhub's parent service
+            monitor = ctx.getServiceParent(service)
+
+            self._deployZenHubWorkerService(
+                ctx, parent=monitor, instances=workers, callLimit=callLimit
+            )
+            log.info(
+                "['%s' hub] Deployed new zenhubworker service", monitor.name
+            )
+            if workers:
+                log.info(
+                    "['%s' hub] Configured zenhubworker service with %s "
+                    "instances", monitor.name, workers
+                )
+            if callLimit:
+                log.info(
+                    "['%s' hub] Configured zenhubworker service to restart "
+                    "after %s calls", monitor.name, callLimit
+                )
+            log.info("Migrated '%s' hub", monitor.name)
+
+        ctx.commit()
+        log.info(
+            "Migrated %s hub%s",
+            len(services), "s" if len(services) != 1 else ""
+        )
+
+    def _limitZenHubInstances(self, service):
+        # Limit ZenHub service to a single instance.
+        service.instances = 1
+        service.instanceLimits.minimum = 1
+        service.instanceLimits.maximum = 1
+        service.instanceLimits.default = 1
+
+    def _removeObsoleteHealthchecks(self, service):
+        # Remove unncessary health checks
+        obsoleteHealthChecks = ("stable_workers", "workers_running")
+        service.healthChecks = filter(
+            lambda hc: hc.name not in obsoleteHealthChecks,
+            service.healthChecks
+        )
+
+    def _updateZenHubConfig(self, service):
+        configFile = next((
+            cf for cf in service.configFiles
+            if cf.name == "/opt/zenoss/etc/zenhub.conf"
+        ), None)
+        if configFile is None:
+            log.warn("No entry in 'ConfigFiles' for zenhub.conf")
+            return None, None
+
+        content = configFile.content
+
+        content, _ = editConfig(workers_entry, content)
+        content, _ = editConfig(worker_priority_entry, content)
+        content, _ = editConfig(call_limit_entry, content)
+        content, _ = editConfig(priority_set, content)
+
+        # Retrieve workers config, if it was set
+        content, groups = editConfig(workers_set, content)
+        workers = groups[0] if groups else None
```

Comment:
Should the new worker count be the number of zenhub instances * worker

Reply:
I chatted with Jake and StevePC and they said yes.  They also recommended that the invalidationworker count should also be adjusted if there were multiple `zenhub` instances.  I added adjusting the invalidationworker count to the migration.